dpctl/tensor/libtensor/include/kernels/sorting/radix_sort.hpp

```diff
@@ -0,0 +1,1918 @@
+        }
+        else {
+            // wg_size * block_size == 64 * 8 * 32 == 16384
+            constexpr std::uint16_t wg_size = ref_wg_size * 8;
+            constexpr std::uint16_t block_size = 32;
+
+            sort_ev = subgroup_radix_sort<_RadixSortKernel, wg_size, block_size,
+                                          radix_bits>{}(
+                exec_q, n_iters, n_to_sort, input_arr, output_arr, proj_op,
+                is_ascending, depends);
+        }
+    }
+    else {
+        constexpr std::uint32_t radix_iters =
+            number_of_buckets_in_type<KeyT>(radix_bits);
+        constexpr std::uint32_t radix_states = std::uint32_t(1) << radix_bits;
+
+        constexpr std::size_t bound_512k = (std::size_t(1) << 19);
+        constexpr std::size_t bound_2m = (std::size_t(1) << 21);
+
+        const auto wg_sz_k = (n_to_sort < bound_512k)  ? 8
+                             : (n_to_sort <= bound_2m) ? 4
+                                                       : 1;
+        const std::size_t wg_size = max_wg_size / wg_sz_k;
+
+        const std::size_t n_segments = (n_to_sort + wg_size - 1) / wg_size;
+
+        // Additional radix_states elements are used for getting local offsets
+        // from count values + no_op flag; 'No operation' flag specifies whether
+        // to skip re-order phase if the all keys are the same (lie in one bin)
+        const std::size_t n_counts =
+            (n_segments + 1) * radix_states + 1 /*no_op flag*/;
+
+        using CountT = std::uint32_t;
+
+        // memory for storing count and offset values
+        CountT *count_ptr =
+            sycl::malloc_device<CountT>(n_iters * n_counts, exec_q);
+        if (nullptr == count_ptr) {
+            throw std::runtime_error("Could not allocate USM-device memory");
+        }
+
+        constexpr std::uint32_t zero_radix_iter{0};
+
+        if constexpr (std::is_same_v<KeyT, bool>) {
+
+            sort_ev = parallel_radix_sort_iteration_step<
+                radix_bits, /*even=*/true>::submit(exec_q, n_iters, n_segments,
+                                                   zero_radix_iter, n_to_sort,
+                                                   input_arr, output_arr,
+                                                   n_counts, count_ptr, proj_op,
+                                                   is_ascending, depends);
+
+            sort_ev = exec_q.submit([=](sycl::handler &cgh) {
+                cgh.depends_on(sort_ev);
+                const sycl::context &ctx = exec_q.get_context();
+
+                using dpctl::tensor::alloc_utils::sycl_free_noexcept;
+                cgh.host_task(
+                    [ctx, count_ptr]() { sycl_free_noexcept(count_ptr, ctx); });
+            });
+
+            return sort_ev;
+        }
+
+        ValueT *tmp_arr =
+            sycl::malloc_device<ValueT>(n_iters * n_to_sort, exec_q);
+        if (nullptr == tmp_arr) {
+            sycl::free(count_ptr, exec_q);
```

Comment:
Actually, I've just realized on re-read: should these `sycl::free` calls be replaced with `sycl_free_noexcept` @oleksandr-pavlyk ?